### ISSUE TITLE
type inheritance not taken into account depending on order of declaration

### ISSUE THREAD

**Issue by gewang** [ccrma/chuck#349]
The following code should work, but does not:
```
public class Foo
{
    BarFactory.make() @=> Bar bar;
}

class Bar
{
    "i am bar" => string name;
}

class BarChild extends Bar
{
    "i extend bar" => name;
}

class BarFactory
{
    fun static BarChild make()
    {
        return new BarChild;
    }
}
```
compiler error:
```
error: cannot assign '@=>' on types 'BarChild' @=> 'Bar'...
[3]     BarFactory.make() @=> Bar bar;
        ^
error:  |- reason: incompatible types for assignment
[3]     BarFactory.make() @=> Bar bar;
```
Cause: class definitions are processed across several phase (scan0, scan1, scan2, type-checking, code-emission). While the new Type defined by the class (e.g., `BarChild`) is created during scan (scan0), the inheritance (`class BarChild extends Bar`) is not processed until the type-checking phase, where it is done in the order it appears in the code. In this case, since `Foo`'s `BarFactory.make() @=> Bar bar;` occurs before `class BarChild extends Bar`, `Bar` and `BarChild` are known but that `BarChild` extends `Bar` is not yet known to the type system.

Possible Remedy:
1) move the class inheritance check earlier, from type-checking potentially into scan1 or scan2 (and leave the class body portion in type-checking)
2) need to check for circular extensions e.g., `A extend B` and `B extends A`

(FYI issue uncovered while addressing #348 cc @nshaheed @heuremh)

**Comment by heuermh** [comment]
See also related issue https://github.com/ccrma/chuck/issues/107

**Comment by gewang** [comment]
addressed in dbcec841135bd9f24c7357ccfb1f50c542443c4b and b6206584314c1b8237264eddb82c85858fcf0698; will be part of 1.5.2.0 release
